spanner/transaction.go

```diff
@@ -952,6 +952,28 @@ func (t *ReadWriteTransaction) begin(ctx context.Context) error {
 	return err
 }
 
+// Commit tries to commit a readwrite transaction to Cloud Spanner. It also
+// returns the commit timestamp for the transactions.
+//
+// This method should be used with client.BeginReadWriteTransaction and only be
+// used if a custom error handling is needed. For normal use cases,
+// client.ReadWriteTransaction should be used because it has robust error
+// handlings and retries.
+func (t *ReadWriteTransaction) Commit(ctx context.Context) (time.Time, error) {
+	var (
+		ts  time.Time
+		err error
+	)
+	ts, err = t.commit(ctx)
+	if err != nil {
+		t.rollback(ctx)
```

Comment:
This should check whether the error was an `Aborted` error, and only execute a `rollback` in case it was **not** an `Aborted` error. Rolling back an aborted transaction is not necessary.

Reply:
If I understand it correctly, e.g., 

```bash
1. BeginReadWriteTransaction
2. Execute statement 1
3. Execute statement 2
4. Execute statement 3
5. Commit // returns an Aborted error
```

Assume that the commit on Step 5 throws an `Aborted` error. This means that the server has cancelled the entire transaction and we don't need to rollback. However, if it is not an `Aborted` error, then we should rollback to actively tell the server that we want to cancel this transaction. 

Also, I would like to confirm the retry logic for an `Aborted` error as well. E.g., 

```bash
1. BeginReadWriteTransaction
2. Execute statement 1
3. Execute statement 2
4. Execute statement 3 // returns an Aborted error
5. Commit
```

The above Step 4 throws an `Aborted` error. If we want to retry this transaction, we can just directly call: 

```bash
2. Execute statement 1
3. Execute statement 2
4. Execute statement 3
5. Commit
```

Is this right? So we don't need to call `rollback` to cancel the previous executed Statement 1 & 2 (they have been aborted by the server).

Reply:
I think if we want to retry the entire transaction after an `Aborted` error, we need to call `BeginReadWriteTransaction` to get a new transaction because the previous txn is already closed. So the retry needs to call the following logic again: 

```bash
1. BeginReadWriteTransaction
2. Execute statement 1
3. Execute statement 2
4. Execute statement 3
5. Commit
```

Reply:
> Assume that the commit on Step 5 throws an `Aborted` error. This means that the server has cancelled the entire transaction and we don't need to rollback. However, if it is not an `Aborted` error, then we should rollback to actively tell the server that we want to cancel this transaction.

Correct.

Reply:
> I think if we want to retry the entire transaction after an `Aborted` error, we need to call `BeginReadWriteTransaction` to get a new transaction because the previous txn is already closed. So the retry needs to call the following logic again:

Correct. If the transaction is aborted, we do not need to call `Rollback`, but we do need to call `BeginReadWriteTransaction` again. The simplest way is to just put the entire transaction in a loop and break out of that loop as long as no error is detected. This is basically what the current ReadWriteTransaction function does.

The tricky part here is of course that the users that want to use statement based transactions probably want to do so because they don't know in advance what all the statements of transaction will be. That again makes it impossible to actually place these in a normal loop. I would still suggest that we put a clear example of how to do that in the documentation of the `BeginReadWriteTransaction` method. That is the entry point for statement based transactions, and should be the first the user should read.

This is how it's documented in Java: https://github.com/googleapis/java-spanner/blob/462839b625e58e235581b8ba10b398e1d222eaaf/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClient.java#L259